### PR TITLE
ERC20Guild Tests

### PR DESCRIPTION
**Pull request by AugustoL**

### CODE REVIEW

File: contracts/erc20guild/ERC20Guild.sol [hunk 1]
```diff
@@ -136,7 +144,10 @@ contract ERC20Guild {
                 .call.value(proposals[proposalId].value[i])(proposals[proposalId].data[i]);
             require(success, "ERC20Guild: Proposal call failed");
         }
-        
+
+        Proposal storage proposal = proposals[proposalId];
```
Comment (AugustoL) [inline contracts/erc20guild/ERC20Guild.sol:148]: Instead of assiging teh storage variable, cant we just assign it to true? `proposals[proposalId].executed = true` it would be more gas efficient

**Comment by AugustoL** [comment]
Few things:
- Test and coverage instructions missing. We should have it in readme.
- Tests missing for each implementation, we should have a folder in tests for erc20guild tests and its implementation, with a file for each erc20guild.

File: contracts/erc20guild/ERC20Guild.sol (hunk 1 above)
Comment (andygray) [reply contracts/erc20guild/ERC20Guild.sol:148]: I need to check whether that is memory or storage. Will adjust and run tests

File: test/ERC20GuildLockable.js [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+require('chai').should();
+
+const ProposalState = {
+    submitted: 0,
+    passed: 1,
+    failed: 2,
+    executed: 3
+};
+
+contract("ERC20GuildLockable", function (accounts) {
```
Comment (AugustoL) [inline test/ERC20GuildLockable.js:22]: We need stronger tests, not just the lock of tokens works, we want to test the positive and negative votes of multiples proposals with locked tokens.

File: test/ERC20GuildPayable.js [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+        walletScheme = await WalletScheme.new();
+        votingMachine = await helpers.setupGenesisProtocol(
+            accounts, orgToken.address, 0, helpers.NULL_ADDRESS
+        );
+    });
+
+    describe("with high gas vote setting (above cost) and standard gas price", function () {
+
+        const VOTE_GAS = new BN('95000'); // 95k - more than enough
+        const MAX_GAS_PRICE = new BN('8000000000'); // 8 gwei
```
Comment (AugustoL) [inline test/ERC20GuildPayable.js:49]: The max gas price should be higher than the default used by the tests, and test the refunds with lower than max, same as max and higher than max, you only need to set the max gas price for the tests once and test with custom gas prices or default gas prices, but it would be better to in this test cases to always set the gas price used in the setVote function.

File: test/ERC20GuildPermissioned.js [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+          canGenericCall: true,
+          canUpgrade: true,
+          canChangeConstraints: true,
+          canRegisterSchemes: true
+        }) ],
+        "metaData"
+      );
+    });
+
+    describe.only("Initialisation", function() {
```
Comment (AugustoL) [inline test/ERC20GuildPermissioned.js:98]: initialiZation

File: test/ERC20GuildPermissioned.js [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        try {
+          await erc20GuildPermissioned.initialize(guildToken.address, 10, 10, 10);
+          assert(false, "ERC20Guild: Only callable by ERC20guild itself when initialized");
+        } catch(error) {
+          helpers.assertVMException(error);
+        }
+      });
+    });
+
+    describe.only("Out of the box permissioned proposals", function() {
```
Comment (AugustoL) [inline test/ERC20GuildPermissioned.js:119]: out of the box? what do you mean with out of the box? please be specific.

File: test/ERC20GuildPermissioned.js [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+          "ERC20GuildPermissioned: Not allowed call"
+        );
+
+        // Check values are still the same
+        (await erc20GuildPermissioned.minimumProposalTime()).should.be.bignumber.equal(minProposalTime);
+        (await erc20GuildPermissioned.votesForExecution()).should.be.bignumber.equal(votesForExecution);
+        (await erc20GuildPermissioned.votesForCreation()).should.be.bignumber.equal(votesForCreation);
+      });
+
+      it("Reverts when trying to get the permissioned guild to call an unauthorised method", async function() {
```
Comment (AugustoL) [inline test/ERC20GuildPermissioned.js:247]: unauithoriZed

File: test/ERC20GuildPermissioned.js [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+        const initializeFunctionEncoded = await new web3.eth.Contract(ERC20GuildPermissioned.abi).methods.initialize(
+          guildToken.address, minProposalTime, votesForExecution, votesForCreation
+        ).encodeABI();
+
+        const txGuild = await erc20GuildPermissioned.createProposal(
+          [erc20GuildPermissioned.address], // I want to invoke a fn on the guild
+          [initializeFunctionEncoded], // The function I want to invoke
+          ['0'], // you sir, get no money.
```
Comment (AugustoL) [inline test/ERC20GuildPermissioned.js:255]: Please remove those comments

File: test/ERC20GuildPermissioned.js (hunk 6 above)
Comment (AugustoL) [inline test/ERC20GuildPermissioned.js:247]: Can we you test it with a more realistic unauthorized method? we have contracts that can be used for this in the contract/test folder.

File: test/ERC20GuildPermissioned.js [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+    daoCreator = await DaoCreator.new(
+      controllerCreator.address,
+    );
+    walletScheme = await WalletScheme.new();
+    votingMachine = await helpers.setupGenesisProtocol(
+      accounts, orgToken.address, 0, helpers.NULL_ADDRESS
+    );
+  });
+
+  describe.only("ERC20GuildPermissioned Core Tests", function() {
```
Comment (AugustoL) [inline test/ERC20GuildPermissioned.js:61]: I would recoomend configure the guild to call an specific contract call in the before each, you have contracts that can be used for it in the contracts/test folder, and then test that it can only execute that call, it cant execute the same call signature on another contract receiver and also that it can de removed with setAllowance.

File: test/ERC20GuildSigned.js [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        daoCreator = await DaoCreator.new(
+            controllerCreator.address, {gas: constants.ARC_GAS_LIMIT}
+        );
+        walletScheme = await WalletScheme.new();
+        votingMachine = await helpers.setupGenesisProtocol(
+            accounts, orgToken.address, 0, helpers.NULL_ADDRESS
+        );
+    });
+
+    describe("ERC20GuildSigned", function () {
```
Comment (AugustoL) [inline test/ERC20GuildSigned.js:45]: Proposal execution and verification of it execution results missing in the tets, also sending signatures form invalid proposals and votes missing, very light tests..

File: test/ERC20GuildPermissioned.js (hunk 4 above)
Comment (andygray) [reply test/ERC20GuildPermissioned.js:98]: We are British but can switch to American spelling...

Comment (vince0656) [reply test/ERC20GuildPermissioned.js]: Fixed

File: test/ERC20GuildPermissioned.js (hunk 5 above)
Comment (vince0656) [reply test/ERC20GuildPermissioned.js]: fixed (hopefully it makes more sense now)

File: test/ERC20GuildPermissioned.js (hunk 6 above)
Comment (vince0656) [reply test/ERC20GuildPermissioned.js]: fixed

File: test/ERC20GuildPermissioned.js (hunk 7 above)
Comment (vince0656) [reply test/ERC20GuildPermissioned.js]: fixed

File: test/ERC20GuildPermissioned.js (hunk 6 above)
Comment (vince0656) [reply test/ERC20GuildPermissioned.js]: fixed

File: contracts/erc20guild/ERC20GuildLockable.sol [hunk 10]
```diff
@@ -73,11 +73,16 @@ contract ERC20GuildLockable is ERC20Guild {
...
+        require(votesOf(msg.sender) >= amount, "ERC20GuildLockable: Unable to release more tokens than locked");
+
+        // FIXME BlockRocket flipped this to < from > - please confirm correct
+        require(tokensLocked[msg.sender].timestamp < block.timestamp, "ERC20GuildLockable: Tokens still locked");
+
+        // FIXME Re-Entrancy? BlockRocket suggest setting amounts before the transfer for safety...
         token.transfer(msg.sender, amount);
         tokensLocked[msg.sender].amount = tokensLocked[msg.sender].amount.sub(amount);
         totalLocked = totalLocked.sub(amount);
+
```
Comment (andygray) [inline contracts/erc20guild/ERC20GuildLockable.sol:85]: https://diligence.consensys.net/blog/2019/09/stop-using-soliditys-transfer-now/

```The simplest way to eliminate reentrancy bugs is to use the checks-effects-interactions pattern. Here’s a classic example of a reentrancy bug:

 1contract Vulnerable {
 2    ...
 3
 4    function withdraw() external {
 5        uint256 amount = balanceOf[msg.sender];
 6        (bool success, ) = msg.sender.call.value(amount)("");
 7        require(success, "Transfer failed.");
 8        balanceOf[msg.sender] = 0;
 9    }
10}
If msg.sender is a smart contract, it has an opportunity on line 6 to call withdraw() again before line 7 happens. In that second call, balanceOf[msg.sender] is still the original amount, so it will be transferred again. This can be repeated as many times as necessary to drain the smart contract.

The idea of the checks-effects-interactions pattern is to make sure that all your interactions (external calls) happen at the end. A typical fix for the above code is as follows:

 1contract Fixed {
 2    ...
 3
 4    function withdraw() external {
 5        uint256 amount = balanceOf[msg.sender];
 6        balanceOf[msg.sender] = 0;
 7        (bool success, ) = msg.sender.call.value(amount)("");
 8        require(success, "Transfer failed.");
 9    }
10}
Notice that in this code, the balance is zeroed out before the transfer, so attempting to make a reentrant call to withdraw() will not benefit an attacker.```

File: contracts/erc20guild/ERC20Guild.sol [hunk 11]
```diff
@@ -212,9 +222,10 @@ contract ERC20Guild {
     /// @dev Get the voting power of multiple addresses
     /// @param accounts The addresses of the token accounts
     function votesOf(address[] memory accounts) public view returns(uint256[] memory) {
-      uint256[] memory votes;
-      for(uint i = 0; i < accounts.length; i ++)
-      votes[i] = votesOf(accounts[i]);
+      uint256[] memory votes = new uint256[](accounts.length);
+      for (uint i = 0; i < accounts.length; i ++) {
```
Comment (andygray) [inline contracts/erc20guild/ERC20Guild.sol:226]: was throwing opcode error so initialised array

File: test/helpers/guild.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    contentHash,
+    extraTime,
+    {from: account}
+  );
+
+  // Return proposal ID
+  return helpers.getValueFromLogs(tx, "proposalId", "ProposalCreated");
+}
+
+export async function useAllVotesOnProposal(guild, proposalId, account) {
```
Comment (AugustoL) [inline test/helpers/guild.js]: what about having a default value of votes to be used to zero as last argument? that would use all votes by default, this would give you more fexibility for the tests.

File: test/helpers/guild.js [hunk 13]
```diff
@@ -1,5 +1,17 @@
 import * as helpers from "./index";
 const ERC20Mock = artifacts.require("ERC20Mock.sol");
+const { BN } = require("@openzeppelin/test-helpers");
+
+export async function createAndSetupGuildToken(accounts) {
```
Comment (AugustoL) [inline test/helpers/guild.js]: It would be better to receive an array of accounts and balances and  create the guidl with that, it will give you us more flexibility for the tests.

File: test/helpers/guild.js [hunk 14]
```diff
@@ -13,6 +16,55 @@ export async function createAndSetupGuildToken(accounts) {
   return guildToken;
 }
 
+export async function createDAO(guild, accounts) {
```
Comment (AugustoL) [inline test/helpers/guild.js:20]: It would be better to also sent another array for rep amounts, to be used for each account.

File: contracts/erc20guild/ERC20GuildLockable.sol (hunk 10 above)
Comment (AugustoL) [reply contracts/erc20guild/ERC20GuildLockable.sol:86]: Great, please leave a comment on the code and I will change it later.

File: test/helpers/guild.js (hunk 12 above)
Comment (andygray) [reply test/helpers/guild.js]: I have added another helper for this purpose

File: test/helpers/guild.js (hunk 14 above)
Comment (andygray) [reply test/helpers/guild.js:20]: `export async function createDAO(guild, accounts, founderToken=  [ 0, 0, 0, 0 ], founderReputation= [10, 10, 10, 10]) {`

File: test/helpers/guild.js (hunk 13 above)
Comment (andygray) [reply test/helpers/guild.js]: Adjusted.

File: test/helpers/guild.js (hunk 12 above)
Comment (andygray) [reply test/helpers/guild.js]: `setXVotesOnProposal`

File: contracts/erc20guild/ERC20GuildLockable.sol (hunk 10 above)
Comment (andygray) [reply contracts/erc20guild/ERC20GuildLockable.sol:86]: Will leave all FIXMEs in the code for you to resolve once merged

File: test/ERC20GuildLockable.js (hunk 2 above)
Comment (andygray) [reply test/ERC20GuildLockable.js]: Done basic positive and negative `use cases` also added test to ensure tokens are locked
